kep/624-disk-io-aware-scheduling/README.md

```diff
@@ -0,0 +1,207 @@
+
+To support the disk IO resource aware scheduling, we add a scheduler plugin that tracks each pod’s disk IO resource need and does accounting of the available disk IO resource on each node when making the scheduling decision.
+
+## Design Consideration
+
+Unlike CPU and memory, a disk’s available IO bandwidth (BW) cannot be calculated by simply subtracting all the running workloads’ requested IO BW from a disk’s total IO BW capacity. A disk’s total IO BW capacity is not a fixed value and changes dynamically upon the characteristics of the workloads running on it, such as the workload’s disk accessing block size and read/write ratio. At different time points, different combinations of workloads could be scheduled to run on a node in the cloud cluster. Therefore a disk’s total IO BW capacity would change dynamically and a disk’s available IO accounting cannot be done in the same way as the CPU/memory accounting.
+
+Since the remaining IO BW capacity of a disk can change dynamically based on the characteristics of the existing workloads, the characteristics of the workload such as IO block size and read/write ratio must be specified in the pod specification together with the disk IO BW requirement. It could happen that some users do not have knowledge of their workloads’ characteristics. In this case, the scheduler plugin will use default values and later get the workloads’ IO block size and read/write ratio through a real-time metrics collector.
+
+The mathematical relationship between the disk’s IO BW capacity and the running workloads’ characteristics is different for different types of disks manufactured by different vendors. There is no “one size fits all” function to model/normalize it. Therefore, the disk IO scheduler design would provide flexibility for vendors to plug in different calculation/normalization models in the form of an IO Driver.
+
+### Goals
+
+- Implement a disk IO aware scheduler plugin which enables the disk IO aware accounting and scheduling
+
+- Define flexible communication APIs between disk IO aware scheduler plugin and the vendor-specific IO Driver
+
+### Non-Goals
+
+- The implementation of the disk IO Driver due to the distinct characteristics of each disk device.
+
+## Proposal
+
+The disk IO aware scheduler plugin would implement the filter, score and reserve hook points of the scheduler framework. At the filter stage, it would obtain each node’s available disk IO capacity from the IO Driver at run time, update the available disk IO capacity of each node in its local cache, and filter out nodes which do not have enough IO capacity from the node candidate list. At the score stage, it would prioritize the node candidates based on a scoring policy, such as most allocated policy. At the reserve stage it would notify the node IO Driver of the new pod so that the IO Driver can start collecting the disk IO metrics of this new pod.
```

Comment:
@Huang-Wei Thanks for your review! Let me try to clarify. The BW a pod demands is a fixed value specified by the user in the annotation of the pod spec. You are right that a node's BW capacity could change dynamically upon different combinations of workloads running on it. But at the time point when the scheduler plugin performs the filtering operation upon scheduling a new pod, the combination of workloads at each node is known and the capacity of each node at that time point can be calculated. Therefore the scheduler plugin can check whether the disk IO capacity of each node based on the new workload combination is larger than the total disk IO requirements of all the pods including the to-be-scheduled pod, and filter out nodes that do not meet the requirement.

Reply:
OK, I see your point. For best-effort workloads, yes, we allow it to run in a non-optimal way and do not apply filtering. But for critical workloads that need a kind of guaranteed disk IO performance (this is the requirement we got from customers like Pinterest), we need to make sure the node has enough disk IO BW for those critical workloads so that those critical workloads do not compete for disk IO resources since when that happens, noisy neighbor problem could arise which may lead to serious performance degradation of those critical workloads.

Reply:
in that case, separate them with two annotations: "hard" and "soft", and only invoke Filter for hard requirement.

Reply:
Yes, we have done that in the annotation. If the user specifies the requested disk IO BW value, it is guaranteed workload, i.e. "hard" requirement. If the user does not specify the requested disk IO BW value, then it is the best effort workload, i.e. "soft".